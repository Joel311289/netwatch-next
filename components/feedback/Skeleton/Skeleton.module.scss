.skeletonWrapper {
  display: block;
  transition: all 0.2s ease-in;
  overflow: hidden;
  cursor: default !important;
  position: relative;

  &.circular {
    border-radius: 50% !important;
  }

  &.heading {
    height: 23px;
  }

  &.text {
    height: 19px;
  }

  &::before {
    content: '';
    display: block;
  }
}

.skeleton {
  height: inherit !important;
  transition: all 0.2s;
  transform: none !important;
  background-color: rgb(var(--rgb-color-skeleton)) !important;
  border-radius: 3px !important;

  .button & {
    border-radius: var(--border-radius-button) !important;
  }

  .paragraph & {
    height: 19px !important;
  }

  .ratio & {
    border-radius: 0 !important;
    position: absolute !important;
    max-width: 100% !important;
  }
}

.skeleton.paragraphLine {
  &:last-child {
    width: 50%;
  }

  & + & {
    margin-top: 2px;
  }
}
