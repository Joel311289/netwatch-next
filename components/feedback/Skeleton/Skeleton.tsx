import { Skeleton as SkeletonMUI } from '@mui/material';
import cn from 'classnames';
import { ElementChildren, ElementHTML, Rectangle } from '@/types';
import { useSizeRatio } from '@/hooks/useSizeRatio';
import styles from '@/components/feedback/Skeleton/Skeleton.module.scss';

type SkeletonVariant = 'circular' | 'heading' | 'text' | 'button' | 'paragraph' | 'default';

export type SkeletonProps = typeof defaultProps &
  ElementHTML &
  ElementChildren &
  Rectangle & {
    ratio?: number;
    variant?: SkeletonVariant;
    numOfLines?: number;
  };

const defaultProps = {
  variant: 'default'
};

const Skeleton = ({
  className,
  children,
  width,
  height,
  ratio,
  variant,
  numOfLines
}: SkeletonProps) => {
  const sizes = useSizeRatio({
    width,
    height,
    ratio: variant === 'circular' ? 1 : ratio
  });

  return (
    <div
      role="progressbar"
      aria-label="skeleton"
      className={cn(styles.skeletonWrapper, styles[variant], ratio && styles.ratio, className)}>
      {variant === 'paragraph' &&
        Array(numOfLines)
          .fill('')
          .map((_, index) => (
            <SkeletonMUI
              key={index}
              className={cn(styles.skeleton, styles.paragraphLine)}
              animation="wave">
              {children}
            </SkeletonMUI>
          ))}

      {variant !== 'paragraph' && (
        <SkeletonMUI className={styles.skeleton} animation="wave">
          {children}
        </SkeletonMUI>
      )}

      <style jsx>{`
        .${styles.skeletonWrapper} {
          width: ${sizes.width || 'auto'};
          min-width: ${sizes.width || 'auto'};
          height: ${sizes.height || 'auto'};
        }
        .${styles.skeletonWrapper}::before {
          padding-top: ${ratio ? `${100 * ratio}%` : '0'};
        }
      `}</style>
    </div>
  );
};

Skeleton.defaultProps = defaultProps;

export default Skeleton;
