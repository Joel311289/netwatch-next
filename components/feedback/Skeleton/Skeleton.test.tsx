import { render, screen } from '@testing-library/react';
import Skeleton from '@/components/feedback/Skeleton/Skeleton';

describe('Tests Skeleton component', () => {
  test('renders component correctly', () => {
    const { container } = render(<Skeleton />);

    expect(container).toMatchSnapshot();
  });

  test('renders with circular variant correctly', () => {
    const width = 200;
    render(<Skeleton variant="circular" width={width} />);

    expect(screen.getByRole('progressbar')).toHaveStyle(`width: ${width}px`);
    expect(screen.getByRole('progressbar')).toHaveStyle(`height: ${width}px`);
  });
});
