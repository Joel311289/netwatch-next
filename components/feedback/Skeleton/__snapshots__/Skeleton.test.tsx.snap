// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Skeleton component renders component correctly 1`] = `
<div>
  <div
    class="skeletonWrapper default"
    role="progressbar"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave skeleton css-1dk1h2u-MuiSkeleton-root"
    />
    <style>
      
        .skeletonWrapper {
          width: auto;
          min-width: auto;
          height: auto;
        }
        .skeletonWrapper::before {
          padding-top: 0;
        }
      
    </style>
  </div>
</div>
`;
